File: src/main/scala/com/xiuhong/flink/applicationdevelopment/streaming/overview/Transformations.scala
package com.xiuhong.flink.applicationdevelopment.streaming.overview

import org.apache.flink.streaming.api.scala._

/**
  * author:xiuhong
  * date:6/12/17
  * time:08:02:22
  */
object Transformations {
  def main(args: Array[String]): Unit = {
    //获取执行环境
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    //读入一个元素，返回转换后的一个元素。一个把输入流转换中的数值翻倍的map function：
    // val num = env.fromElements(1, 2, 3, 4)
    // val result = num.map(_ * 2)

    //读入一个元素，返回转换后的0个、1个或者多个元素。一个将句子切分成单词的flatmap function：
    //    val text = env.fromElements("i love you yexing")
    //    val words = text.flatMap { str => str.split(" ") }
    //    words.print()

    //读入的每个元素执行boolean函数，并保留返回true的元素。一个过滤掉零值的filter：
    //    val num = env.fromElements(1, 2, 3, 4)
    //    val result = num.filter(x => x > 2)
    //    result.print()
    //    env.execute()


    //逻辑上将流分区为不相交的分区，每个分区包含相同key的元素。
    // 在内部通过hash分区来实现。关于如何指定分区的keys请参阅keys。
    // 该transformation返回一个KeyedDataStream。
    //    val text = env.fromElements("i love you yexing i love you jeffdean")
    //    val words = text.flatMap { str => str.split(" ") }
    //      .map(x => (x, 1))
    //      .keyBy(0)
    //    words.print()

  }
}

File: src/main/java/com/xiuhong/flink/dataartisans/datastreamapi/one/Transformations.java
package com.xiuhong.flink.dataartisans.datastreamapi.one;

import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.api.java.tuple.Tuple;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

import java.util.ArrayList;
import java.util.List;

/**
 * author:xiuhong
 * date:6/9/17
 * time:09:20:06
 */
public class Transformations {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        DataStream<Person> p = env.fromElements(new Person(1, "Bob"));
        p.keyBy("id");

        p.print();
        env.execute();

    }
}

class Person {
    public int id;
    public String name;

    public Person() {
    }

    ;

    public Person(int id, String name) {
        this.id = id;
        this.name = name;

    }

    ;
}